index/scorch/segment/zap/merge.go

```diff
@@ -361,14 +349,30 @@ func persistMergedRest(segments []*SegmentBase, dropsIn []*roaring.Bitmap,
 
 		// update the field doc values
 		fdvEncoder := newChunkedContentCoder(uint64(chunkFactor), newSegDocCount-1, w, true)
-		for docNum, docTerms := range docTermMap {
-			if len(docTerms) > 0 {
-				err = fdvEncoder.Add(uint64(docNum), docTerms)
+
+		for segmentI, segment := range segmentsInFocus {
+			fieldIDPlus1 := uint16(segment.fieldsMap[fieldName])
+			if dvIter, exists := segment.fieldDvReaders[fieldIDPlus1-1]; exists &&
+				dvIter != nil {
+				i := 0
+				err = dvIter.iterateAllDocValues(segment, func(docNum uint64, terms []byte) error {
+					if newDocNums[segmentI][i] == docDropped {
+						i++
+						return nil
+					}
+					err := fdvEncoder.Add(newDocNums[segmentI][i], terms)
```

Comment:
using 'i' here feels wrong?  That is, using the callback param 'docNum' here instead of 'i' feels more right to me?

So, more like... newDocUms[segmentI][docNum] 

Thoughts?

Reply:
Yes agreed, docNums will align with index positions of newDocNums[seg..].